### PR TITLE
#22993 adding permissions on save draft

### PR DESCRIPTION
**Pull request by jdotcms**
Starting to include the permissions on the save draft

### CODE REVIEW

File: dotCMS/src/main/java/com/dotcms/rest/api/v1/workflow/WorkflowResource.java [hunk 1]
```diff
@@ -1602,20 +1605,26 @@ private Response fireAction(final HttpServletRequest request,
...
                 entry.getValue().forEach(roleId -> permissions.add(
-                        new Permission(null, roleId, entry.getKey().getType())));
+                        new Permission(null, this.mapRoleId(roleId), entry.getKey().getType())));
             }
 
             formBuilder.permissions(permissions);
         }
     }
 
+    private String mapRoleId (final String roleIdOrKey) {
```
Comment (nollymar) [inline dotCMS/src/main/java/com/dotcms/rest/api/v1/workflow/WorkflowResource.java:1622]: we would need a postman or IT to validate that:
- we can support roleId/roleKey
- if the roleId/roleKey is invalid, we ignore it and continue with the remaining permissions

File: dotCMS/src/main/java/com/dotcms/rest/api/v1/workflow/WorkflowResource.java [hunk 2]
```diff
@@ -1602,20 +1605,26 @@ private Response fireAction(final HttpServletRequest request,
     private void processPermissions(final FireActionForm fireActionForm,
                                     final ContentletDependencies.Builder formBuilder) {
 
-        if (UtilMethods.isSet(fireActionForm.getIndividualPermissions())) {
+        if (null != fireActionForm.getIndividualPermissions()) {
```
Comment (nollymar) [inline dotCMS/src/main/java/com/dotcms/rest/api/v1/workflow/WorkflowResource.java:1608]: did you implement this https://github.com/dotCMS/core/issues/22993#issuecomment-1248621401?

File: dotCMS/src/main/java/com/dotmarketing/portlets/workflows/actionlet/SaveContentAsDraftActionlet.java [hunk 3]
```diff
@@ -78,15 +78,16 @@ public void executeAction(final WorkflowProcessor processor,
...
-			final List< Permission > permissions    =
-					this.permissionAPI.getPermissions(contentlet, false, true);
+			List<Category > categories              = null;
 
 			Logger.debug(this,
 					"Saving the content as draft of the contentlet: " + contentlet.getIdentifier());
 
 			contentlet.setProperty(Contentlet.WORKFLOW_IN_PROGRESS, Boolean.TRUE);
 			final ContentletDependencies contentletDependencies = processor.getContentletDependencies();
+			final List< Permission > permissions    = null != contentletDependencies && null != contentletDependencies.getPermissions()?
```
Comment (nollymar) [inline dotCMS/src/main/java/com/dotmarketing/portlets/workflows/actionlet/SaveContentAsDraftActionlet.java:88]: I don't remember if we agreed with @wezell to support other actions, not just Save as Draft, do you?

File: dotCMS/src/main/java/com/dotcms/content/elasticsearch/business/ESContentletAPIImpl.java [hunk 4]
```diff
@@ -8808,20 +8831,40 @@ public Contentlet checkin(Contentlet contentlet, ContentletRelationships content
...
-                        .collect(Collectors.toList()), contentletReturned, user, respectFrontendRoles);
+
+                    selectedPermissions.stream()
+                            .map(permission -> new Permission(contentlet.getPermissionId(), permission.getRoleId(), permission.getPermission()))
+                            .forEach(permission -> {
+                                try {
+                                    Logger.debug(this, ()-> "Adding permission: " + permission);
+                                    this.permissionAPI.save(permission, contentlet, user, respectFrontendRoles);
+                                } catch (Exception e) {
+                                    Logger.error(this, e.getMessage());
```
Comment (nollymar) [inline dotCMS/src/main/java/com/dotcms/content/elasticsearch/business/ESContentletAPIImpl.java:8864]: I would add more details on this error message, for example: contentletId, roleId and permissionId

File: dotCMS/src/integration-test/java/com/dotmarketing/portlets/workflows/actionlet/SaveContentAsDraftActionletTest.java [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+        permissions.add(new Permission(null, role.get().getId(), PermissionAPI.PERMISSION_READ));
+        final ContentletDependencies contentletDependencies =
+                new ContentletDependencies.Builder()
+                .modUser(APILocator.systemUser())
+                .permissions(permissions)
+                .build();
+        final WorkflowProcessor processor = new WorkflowProcessor(contentlet, APILocator.systemUser());
+        processor.setContentletDependencies(contentletDependencies);
+        new SaveContentAsDraftActionlet().executeAction(processor, null);
+        DateUtil.sleep(DateUtil.SECOND_MILLIS*2);
```
Comment (nollymar) [inline dotCMS/src/integration-test/java/com/dotmarketing/portlets/workflows/actionlet/SaveContentAsDraftActionletTest.java:67]: why is this sleep needed?

**Review by nollymar — CHANGES_REQUESTED** [review]

**Comment by nollymar** [comment]
Some unit tests aren't compiling

File: dotCMS/src/main/java/com/dotcms/content/elasticsearch/business/ESContentletAPIImpl.java (hunk 4 above)
Comment (wezell) [reply dotCMS/src/main/java/com/dotcms/content/elasticsearch/business/ESContentletAPIImpl.java:8864]: agree

File: dotCMS/src/main/java/com/dotcms/content/elasticsearch/business/ESContentletAPIImpl.java [hunk 6]
```diff
@@ -5084,7 +5095,7 @@ private boolean addOrUpdateContentletIdentifier(final Contentlet contentlet, fin
             identifier = identifierAPI.find(contentlet);
 
             if (null == identifier || !UtilMethods.isSet(identifier.getId())){
-                throw new DotDataException(String.format("Contentlet with ID '%s' has not been found.", contentlet.getIdentifier()));
+                throw new DotDataException(new DoesNotExistException(String.format("Contentlet with ID '%s' has not been found.", contentlet.getIdentifier())));
```
[the remark this answers is not in the document]
Comment (wezell) [reply dotCMS/src/main/java/com/dotcms/content/elasticsearch/business/ESContentletAPIImpl.java:5098]: did it?

File: dotCMS/src/main/java/com/dotcms/rest/api/v1/workflow/WorkflowResource.java (hunk 2 above)
Comment (jdotcms) [reply dotCMS/src/main/java/com/dotcms/rest/api/v1/workflow/WorkflowResource.java:1608]: yes

File: dotCMS/src/main/java/com/dotcms/rest/api/v1/workflow/WorkflowResource.java (hunk 1 above)
Comment (jdotcms) [reply dotCMS/src/main/java/com/dotcms/rest/api/v1/workflow/WorkflowResource.java]: Done

File: dotCMS/src/main/java/com/dotmarketing/portlets/workflows/actionlet/SaveContentAsDraftActionlet.java (hunk 3 above)
Comment (jdotcms) [reply dotCMS/src/main/java/com/dotmarketing/portlets/workflows/actionlet/SaveContentAsDraftActionlet.java:88]: the method is one the ESContentletAPI, however we need to do this specific work here to make it work

File: dotCMS/src/integration-test/java/com/dotmarketing/portlets/workflows/actionlet/SaveContentAsDraftActionletTest.java (hunk 5 above)
Comment (jdotcms) [reply dotCMS/src/integration-test/java/com/dotmarketing/portlets/workflows/actionlet/SaveContentAsDraftActionletTest.java:67]: b.c the permission are stored on a commit listener which is a separated thread

File: dotCMS/src/main/java/com/dotcms/content/elasticsearch/business/ESContentletAPIImpl.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (jdotcms) [reply dotCMS/src/main/java/com/dotcms/content/elasticsearch/business/ESContentletAPIImpl.java:5155]: Nope it is just wrapping but keeping the same exception

**Pull request closed without merging**